src/models/Ruleserver.ts

```diff
@@ -0,0 +1,290 @@
+    // First key is the room id and the second is the event id.
+    private rulesByEvent: Map<string, Map<string, EventRules>> = new Map();
+    // A unique identifier for this server instance that is given to each response so we can tell if the token
+    // was issued by this server or not. This is important for when Mjolnir has been restarted
+    // but the client consuming the rules hasn't been and we need to tell the client we have rebuilt all of the rules (via `reset` in the response).
+    private serverId: string = crypto.randomUUID();
+    private currentToken = 0;
+    private readonly banListUpdateListener = this.update.bind(this);
+    /**
+     * The token is used to seperate EventRules from each other based on when they were added.
+     * The lower the token, the longer a rule has been tracked for (relative to other rules in this Ruleserver).
+     * The token is incremented before adding new rules to be served.
+     */
+    private nextToken(): void {
+        this.currentToken += 1;
+        this.stoppedRules.push([]);
+        this.rulesByToken.push([]);
+    }
+
+    /**
+     * Get a combination of the serverId and currentToken to give to the client.
+     */
+    private get since(): string {
+        return `${this.serverId}::${this.currentToken}`;
+    }
+
+    /**
+     * Get the `EventRules` object for this Matrix event.
+     * @param roomId The room the event came from.
+     * @param eventId The id of the event.
+     * @returns The `EventRules` object describing which rules have been created based on the policy the event represents.
+     */
+    private getEventRules(roomId: string, eventId: string): EventRules|undefined {
+        return this.rulesByEvent.get(roomId)?.get(eventId);
+    }
+
+    /**
+     * Add the EventRule to be served by the ruleserver at the current token.
+     * @param eventRules Add rules for an associated policy room event. (e.g. m.policy.rule.user).
+     */
+    private addEventRules(eventRules: EventRules): void {
+        const [roomId, eventId] = [eventRules.roomId, eventRules.eventId];
+        if (this.rulesByEvent.get(roomId)?.has(eventId)) {
+            throw new Error(`There is already an entry in the RuleServer for rules created from the event ${eventId}.`);
+        }
+        const roomTable = this.rulesByEvent.get(roomId);
+        if (roomTable) {
+            roomTable.set(eventId, eventRules);
+        } else {
+            this.rulesByEvent.set(roomId, new Map().set(eventId, eventRules));
+        }
+        this.rulesByToken[eventRules.token].push(eventRules);
+    }
+
+    /**
+     * Stop serving the rules from this policy rule.
+     * @param eventRules The EventRules to stop serving from the ruleserver.
+     */
+    private stopEventRules(eventRules: EventRules): void {
+        const eventId = eventRules.eventId;
+        this.rulesByEvent.get(eventRules.roomId)?.delete(eventId);
+        const index = this.rulesByToken[eventRules.token].indexOf(eventRules);
```

Comment:
It might not be actually, we don't make that guarantee and it's very likely that the first (`mjolnir.banLists.length`) rows will have a majority of the rules as most will be added when Mjolnir first starts up and updates each banList for the first time. We could guard this to make sure there will never be more than `n` rules per token, but I'm not sure what would make a good number for `n`.